python/pyarrow/parquet.py

```
@@ -36,6 +38,253 @@
+            # datetime is a subclass of date which is why we need this double check
+            if isinstance(value, datetime.date):
+                value = np.datetime64(value, 'ns')
+            elif isinstance(value, list) and isinstance(value[0], datetime.date):
+                value = [np.datetime64(value, 'ns') for val in value]
+            ser = df[column].values
+            if op == "==":
+                inner_indexer = (ser == value) & inner_indexer
+            elif op == "!=":
+                inner_indexer = (ser != value) & inner_indexer
+            elif op == "<=":
+                inner_indexer = (ser <= value) & inner_indexer
+            elif op == ">=":
+                inner_indexer = (ser >= value) & inner_indexer
+            elif op == "<":
+                inner_indexer = (ser < value) & inner_indexer
+            elif op == ">":
+                inner_indexer = (ser > value) & inner_indexer
+            elif op == "in":
+                inner_indexer = (np.isin(ser, value)) & inner_indexer
+            else:
+                raise NotImplementedError("op not supported")
+        indexer = inner_indexer | indexer
+    return df[indexer]
+
+
+def _check_contains_null(val):
+    if isinstance(val, six.binary_type):
+        for byte in val:
+            if isinstance(byte, six.binary_type):
+                compare_to = chr(0)
+            else:
+                compare_to = 0
+            if byte == compare_to:
+                return True
+    return False
+
+
+def check_filters(filters):
+    """
+    Check if filters are well-formed.
+    """
+    if filters is not None:
+        if len(filters) == 0 or any(len(f) == 0 for f in filters):
+            raise ValueError("Malformed filters")
+        if isinstance(filters[0][0], six.string_types):
+            # We have encountered the situation where we have one nesting level too few:
+            # We have [(,,), ..] instead of [[(,,), ..]]
+            filters = [filters]
```

This automatic fix isn't my favorite though I see that it is necessary to not have a braking change in the api for  ParquetDataset (with the filters argument). Perhaps though it would be better to throw an error here and have this fix in that specific case instead of allowing a wrong nesting level in all cases.